### PR TITLE
Assignment Session 7 (Filter)

### PR DESCRIPTION
**Pull request by jhnwendell02**
Hi Sir @ranton @jegtugado @juliuscnicolas @kevinatgithub here is my homework for session 7. Thank you :)

### CODE REVIEW

File: ASPNetCoreMastersTodoList.Api/Controllers/ItemsController.cs [hunk 1]
```diff
@@ -43,13 +45,15 @@ public IActionResult Post([FromBody] ItemCreateBindingModel itemCreateModel)
...
         [HttpPut("{itemId}")]
+        [EnsureItemIdExistFilterAttribute]
         public IActionResult Put(int itemId, [FromBody] ItemUpdateBindingModel itemUpdateModel)
         {
             ItemDTO requestData = new ItemDTO() { ItemId = itemUpdateModel.ItemId, Text = itemUpdateModel.Name};
             _service.Update(requestData);
             return Ok(requestData);
         }
         [HttpDelete("{itemId}")]
+        [EnsureItemIdExistFilterAttribute]
```
Comment (jegtugado) [inline ASPNetCoreMastersTodoList.Api/Controllers/ItemsController.cs:56]: Hi sir please apply the filter on the controller scope

File: ASPNetCoreMastersTodoList.Api/Filters/EnsureItemIdExistFilter.cs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        {
+            _service = service;
+        }
+        public void OnActionExecuted(ActionExecutedContext context) {}
+
+        public void OnActionExecuting(ActionExecutingContext context)
+        {
+            var itemId = (int)context.ActionArguments["itemId"];
+            var isExist = _service.Get(itemId);
+            if (isExist == null)
```
Comment (jegtugado) [inline ASPNetCoreMastersTodoList.Api/Filters/EnsureItemIdExistFilter.cs:24]: please handle scenarios where the action doesn't need to ensure an item exists (no itemId parameter)

File: ASPNetCoreMastersTodoList.Api/Filters/ExecutionTimeFilter.cs [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+            context.HttpContext.Items["StartTime"] = DateTime.UtcNow;
+            DateTime startTime = (DateTime)context.HttpContext.Items["StartTime"];
+            Console.WriteLine("Total Elapsed Time for Request is : " + (DateTime.UtcNow - startTime).TotalMilliseconds);
+            Console.WriteLine("Request Ended.");
+        }
+
+        public void OnResourceExecuting(ResourceExecutingContext context)
+        {
+            Console.WriteLine("Request started in " + context.ActionDescriptor.DisplayName.ToString());
+            context.HttpContext.Items["StartTime"] = DateTime.UtcNow;
```
Comment (jegtugado) [inline ASPNetCoreMastersTodoList.Api/Filters/ExecutionTimeFilter.cs:23]: excellent approach to prevent race conditions!

Comment (jegtugado) [reply ASPNetCoreMastersTodoList.Api/Filters/ExecutionTimeFilter.cs:23]: next time, we recommend using a StopWatch for easier time elapsed tracking

**Review by jegtugado — CHANGES_REQUESTED** [review]
Hi sir. minor change needed

File: ASPNetCoreMastersTodoList.Api/Filters/ExecutionTimeFilter.cs [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+using System.Threading.Tasks;
+
+namespace ASPNetCoreMastersTodoList.Api.Filters
+{
+    public class ExecutionTimeFilter : IResourceFilter
+    {
+        public void OnResourceExecuted(ResourceExecutedContext context)
+        {
+
+            context.HttpContext.Items["StartTime"] = DateTime.UtcNow;
```
Comment (kevinatgithub) [inline ASPNetCoreMastersTodoList.Api/Filters/ExecutionTimeFilter.cs:14]: May I ask, why are you assigning value to StartTime again here? Thanks

**Review by kevinatgithub — APPROVED** [review]

**Pull request merged**